### PR TITLE
APEXMALHAR-2076 #resolve #comment add AbstractTupleUniqueExactlyOnceK…

### PR DESCRIPTION
**Pull request by brightchen**
…afkaOutputOperator

### CODE REVIEW

**Comment by brightchen** [comment]
@siyuanh Could you please review

File: contrib/src/main/java/com/datatorrent/contrib/kafka/AbstractTupleUniqueExactlyOnceKafkaOutputOperator.java [hunk 1]
```diff
@@ -0,0 +1,610 @@
...
+
+import kafka.api.OffsetRequest;
+import kafka.javaapi.consumer.SimpleConsumer;
+import kafka.javaapi.producer.Producer;
+import kafka.producer.KeyedMessage;
+import kafka.producer.ProducerConfig;
+import kafka.serializer.StringDecoder;
+
+/**
+ * Assumptions: - assume the value of incoming tuples are not duplicate(at least
```
Comment (siyuanh) [inline contrib/src/main/java/com/datatorrent/contrib/kafka/AbstractTupleUniqueExactlyOnceKafkaOutputOperator.java]: Please add html tags to make it more readable

File: contrib/src/main/java/com/datatorrent/contrib/kafka/AbstractTupleUniqueExactlyOnceKafkaOutputOperator.java [hunk 2]
```diff
@@ -0,0 +1,610 @@
...
+   * the recovery window id If all operator partitions crashed, should use the
+   * minimal last window id as the recovery window id, as the data may go to the
+   * other partitions. But as the operator can't distinguish which is the case.
+   * use the most general one.
+   */
+  protected transient long minRecoveryWindowId = -2;
+  protected transient long maxRecoveryWindowId = -2;
+
+  /**
+   * A map from Kafka partition id to lastMessages writtten to this kafka
```
Comment (siyuanh) [inline contrib/src/main/java/com/datatorrent/contrib/kafka/AbstractTupleUniqueExactlyOnceKafkaOutputOperator.java]: Typo

**Comment by brightchen** [comment]
@siyuanh changed comments and typo, please review

**Comment by brightchen** [comment]
@siyuanh Could you please review

File: contrib/src/main/java/com/datatorrent/contrib/kafka/KafkaMetadataUtil.java [hunk 3]
```diff
@@ -104,14 +104,14 @@
...
+
+
   public static Set<String> getBrokers(Set<String> zkHost){
-    
+
     ZkClient zkclient = new ZkClient(StringUtils.join(zkHost, ',') ,30000, 30000, ZKStringSerializer$.MODULE$);
     Set<String> brokerHosts = new HashSet<String>();
     for (Broker b : JavaConversions.asJavaIterable(ZkUtils.getAllBrokersInCluster(zkclient))) {
-      brokerHosts.add(b.getConnectionString());
+      brokerHosts.add(b.connectionString());
```
Comment (siyuanh) [inline contrib/src/main/java/com/datatorrent/contrib/kafka/KafkaMetadataUtil.java:119]: Why this line of change?

File: contrib/src/main/java/com/datatorrent/contrib/kafka/AbstractTupleUniqueExactlyOnceKafkaOutputOperator.java [hunk 4]
```diff
@@ -0,0 +1,612 @@
...
+ * @displayName Abstract Tuple Unique Exactly Once Kafka Output
+ * @category Messaging
+ * @tags output operator
+ */
+@InterfaceStability.Evolving
+public abstract class AbstractTupleUniqueExactlyOnceKafkaOutputOperator<T, K, V>
+    extends AbstractKafkaOutputOperator<K, V>
+{
+  public static final String DEFAULT_CONTROL_TOPIC = "ControlTopic";
+  protected transient int partitionNum = 1;
```
Comment (siyuanh) [inline contrib/src/main/java/com/datatorrent/contrib/kafka/AbstractTupleUniqueExactlyOnceKafkaOutputOperator.java]: Make properties private when you can?  Where is this partitionNum used?

File: contrib/src/main/java/com/datatorrent/contrib/kafka/AbstractTupleUniqueExactlyOnceKafkaOutputOperator.java [hunk 5]
```diff
@@ -0,0 +1,612 @@
...
+
+  //The control info includes the time, use this time to track the head of control info we care.
+  protected int controlInfoTrackBackTime = 120000;
+
+  /**
+   * max number of offset need to check
+   */
+  protected int maxNumOffsetsOfControl = 1000;
+
+  protected String controlProducerProperties;
```
Comment (siyuanh) [inline contrib/src/main/java/com/datatorrent/contrib/kafka/AbstractTupleUniqueExactlyOnceKafkaOutputOperator.java]: We should hide the controlProducerProperties?

File: contrib/src/main/java/com/datatorrent/contrib/kafka/AbstractTupleUniqueExactlyOnceKafkaOutputOperator.java [hunk 6]
```diff
@@ -0,0 +1,612 @@
...
+
+  /**
+   * A map from Kafka partition id to lastMessages written to this kafka
+   * partition. This information was loaded depends on the
+   * RecoveryControlInfo.kafkaPartitionIdToOffset
+   */
+  protected transient Map<Integer, List<Pair<byte[], byte[]>>> partitionToLastMsgs = Maps.newHashMap();
+
+  /**
+   * The messages are assume to written to the kafka partition decided by
```
Comment (siyuanh) [inline contrib/src/main/java/com/datatorrent/contrib/kafka/AbstractTupleUniqueExactlyOnceKafkaOutputOperator.java]: typo

File: contrib/src/main/java/com/datatorrent/contrib/kafka/AbstractTupleUniqueExactlyOnceKafkaOutputOperator.java [hunk 7]
```diff
@@ -0,0 +1,612 @@
...
+
+  private static final Logger logger = LoggerFactory.getLogger(AbstractExactlyOnceKafkaOutputOperator.class);
+
+  /**
+   * This class used to keep the recovery information
+   *
+   */
+  protected static class RecoveryControlInfo
+  {
+    protected static final String SEPERATOR = "#";
```
Comment (siyuanh) [inline contrib/src/main/java/com/datatorrent/contrib/kafka/AbstractTupleUniqueExactlyOnceKafkaOutputOperator.java:567]: Again this should be private as well

File: contrib/src/main/java/com/datatorrent/contrib/kafka/KafkaUtil.java [hunk 8]
```diff
@@ -0,0 +1,358 @@
...
+
+    if (tm == null) {
+      throw new RuntimeException("Failed to retrieve topic metadata");
+    }
+
+    for (PartitionMetadata pm : tm.partitionsMetadata()) {
+      SimpleConsumer consumer = null;
+      try {
+        String clientName = KafkaMetadataUtil.getClientName(clientNamePrefix, tm.topic(), pm.partitionId());
+        consumer = createSimpleConsumer(clientName, tm.topic(), pm);
```
Comment (siyuanh) [inline contrib/src/main/java/com/datatorrent/contrib/kafka/KafkaUtil.java]: Can we reuse consumer if it's reading from same broker?

File: contrib/src/main/java/com/datatorrent/contrib/kafka/AbstractTupleUniqueExactlyOnceKafkaOutputOperator.java (hunk 7 above)
Comment (brightchen) [reply contrib/src/main/java/com/datatorrent/contrib/kafka/AbstractTupleUniqueExactlyOnceKafkaOutputOperator.java:567]: The sub classes might need to use it( for example when over write the logic of AbstractTupleUniqueExactlyOnceKafkaOutputOperator.RecoveryControlInfo.fromString(String) ). modified by private would have a lot of pain in this case.

**Comment by brightchen** [comment]
@siyuanh I applied most change. Please review

**Pull request closed without merging**